File: src/services/incidentService.ts
import type { SourceAdapter } from "@/adapters/types";
import type { Incident, SourceHealth } from "@/types/incident";
import { detectChanges, type ChangeSet } from "./changeDetection";

/**
 * The server-side heart: owns the active set + history (in memory in phase 1;
 * PostgreSQL behind the same interface in phase 2), runs the poll loop against
 * ONE adapter (browsers never poll upstream), detects changes, tracks source
 * health honestly, and fans changes out to subscribers (SSE in phase 4).
 */
export interface ServiceOptions {
  adapter: SourceAdapter;
  pollIntervalMs: number;
  /** Consider the source "delayed" after this many ms without a successful poll. */
  staleAfterMs?: number;
  historyLimit?: number;
  now?: () => Date;
  setTimer?: (fn: () => void, ms: number) => unknown;
  clearTimer?: (t: unknown) => void;
}

export type ChangeListener = (changes: ChangeSet, health: SourceHealth) => void;

export class IncidentService {
  private active = new Map<string, Incident>();
  private history: Incident[] = []; // closed incidents, newest first
  private listeners = new Set<ChangeListener>();
  private timer: unknown = null;
  private inFlight = false;
  private stopped = true;
  private failures = 0;
  private health: SourceHealth;
  private readonly now: () => Date;
  private readonly setTimer: (fn: () => void, ms: number) => unknown;
  private readonly clearTimer: (t: unknown) => void;
  readonly opts: Required<Pick<ServiceOptions, "pollIntervalMs" | "staleAfterMs" | "historyLimit">> & { adapter: SourceAdapter };
  /** Most recent raw sample + parse report for the admin page. */
  lastRawSample: unknown = undefined;
  lastParseErrors: { message: string; record?: string }[] = [];

  constructor(o: ServiceOptions) {
    this.opts = { adapter: o.adapter, pollIntervalMs: o.pollIntervalMs, staleAfterMs: o.staleAfterMs ?? o.pollIntervalMs * 4, historyLimit: o.historyLimit ?? 5000 };
    this.now = o.now ?? (() => new Date());
    this.setTimer = o.setTimer ?? ((fn, ms) => setTimeout(fn, ms));
    this.clearTimer = o.clearTimer ?? ((t) => clearTimeout(t as ReturnType<typeof setTimeout>));
    this.health = { adapter: o.adapter.name, state: "starting", consecutiveFailures: 0, activeCount: 0 };
  }

  start(): void {
    if (!this.stopped) return;
    this.stopped = false;
    void this.pollOnce().then(() => this.schedule(this.opts.pollIntervalMs));
  }

  stop(): void {
    this.stopped = true;
    if (this.timer !== null) this.clearTimer(this.timer);
    this.timer = null;
  }

  private schedule(ms: number): void {
    if (this.stopped) return;
    this.timer = this.setTimer(() => {
      void this.pollOnce().then(() => this.schedule(this.nextDelay()));
    }, ms);
  }

  /** Exponential backoff on failure, capped at 8x the interval. */
  private nextDelay(): number {
    const base = this.opts.pollIntervalMs;
    return this.failures === 0 ? base : Math.min(base * 2 ** this.failures, base * 8);
  }

  /** One poll. Never overlaps: a second call while one is in flight is a no-op. */
  async pollOnce(): Promise<ChangeSet | null> {
    if (this.inFlight) return null;
    this.inFlight = true;
    const started = this.now();
    try {
      const result = await this.opts.adapter.fetchActive();
      const changes = this.apply(result.incidents, started);
      this.failures = 0;
      this.lastRawSample = result.rawSample;
      this.lastParseErrors = result.report.errors;
      this.health = {
        ...this.health, state: "connected", lastPollAt: started, lastSuccessAt: started, lastLatencyMs: result.report.latencyMs,
        consecutiveFailures: 0, lastError: undefined, activeCount: this.active.size,
        lastIncidentUpdateAt: changes.added.length + changes.changed.length + changes.closed.length > 0 ? started : this.health.lastIncidentUpdateAt,
      };
      if (changes.added.length + changes.changed.length + changes.closed.length > 0) this.emit(changes);
      return changes;
    } catch (err) {
      this.failures++;
      const msg = err instanceof Error ? err.message : String(err);
      const stale = !this.health.lastSuccessAt || started.getTime() - this.health.lastSuccessAt.getTime() > this.opts.staleAfterMs;
      this.health = { ...this.health, state: stale ? "down" : "delayed", lastPollAt: started, consecutiveFailures: this.failures, lastError: msg };
      return null;
    } finally {
      this.inFlight = false;
    }
  }

  private apply(incidents: Incident[], now: Date): ChangeSet {
    const changes = detectChanges(this.active, incidents, now);
    for (const i of [...changes.added, ...changes.changed]) this.active.set(i.id, i);
    for (const i of changes.closed) {
      this.active.delete(i.id);
      this.history.unshift(i);
    }
    // An upstream may also report closed incidents explicitly in its active list.
    for (const i of incidents) {
      if (i.status === "closed" && this.active.has(i.id)) {
        this.active.delete(i.id);
        this.history.unshift(i);
      }
    }
    if (this.history.length > this.opts.historyLimit) this.history.length = this.opts.historyLimit;
    return changes;
  }

  private emit(changes: ChangeSet): void {
    for (const l of this.listeners) l(changes, this.getHealth());
  }

  subscribe(l: ChangeListener): () => void {
    this.listeners.add(l);
    return () => this.listeners.delete(l);
  }

  getActive(): Incident[] {
    return [...this.active.values()].sort((a, b) => b.dispatchedAt.getTime() - a.dispatchedAt.getTime());
  }

  getHistory(): Incident[] {
    return [...this.history];
  }

  getById(id: string): Incident | undefined {
    return this.active.get(id) ?? this.history.find((i) => i.id === id);
  }

  /** Health with staleness re-evaluated at read time, so "connected" cannot go stale silently. */
  getHealth(): SourceHealth {
    const h = { ...this.health, activeCount: this.active.size };
    const last = h.lastSuccessAt?.getTime();
    if (h.state === "connected" && last !== undefined && this.now().getTime() - last > this.opts.staleAfterMs) h.state = "delayed";
    return h;
  }
}

File: api/mcfrs_api/services/incident_service.py
"""The server-side heart: owns the active set, runs the poll loop against ONE
adapter (browsers never poll upstream), detects changes, tracks source health
honestly, persists through the repository, and fans changes out to subscribers
(SSE)."""

from __future__ import annotations

import asyncio
import logging
from collections.abc import Awaitable, Callable
from datetime import UTC, datetime, timedelta

from mcfrs_api.adapters.base import SourceAdapter
from mcfrs_api.db.repository import HistoryQuery, IncidentRepository, PollRecord
from mcfrs_api.models import Incident, IncidentStatus, SourceHealth, SourceState
from mcfrs_api.services.change_detection import ChangeSet, detect_changes

log = logging.getLogger(__name__)
Listener = Callable[[ChangeSet, SourceHealth], Awaitable[None] | None]


class IncidentService:
    def __init__(
        self,
        adapter: SourceAdapter,
        repository: IncidentRepository,
        poll_interval_s: float,
        stale_after_s: float | None = None,
        now: Callable[[], datetime] | None = None,
    ) -> None:
        self.adapter = adapter
        self.repo = repository
        self.poll_interval_s = poll_interval_s
        self.stale_after_s = stale_after_s if stale_after_s is not None else poll_interval_s * 4
        self._now = now or (lambda: datetime.now(UTC))
        self._active: dict[str, Incident] = {}
        self._listeners: set[Listener] = set()
        self._lock = asyncio.Lock()
        self._task: asyncio.Task[None] | None = None
        self._failures = 0
        self._health = SourceHealth(adapter=adapter.name)
        self.last_raw_sample: object = None
        self.last_parse_errors: list[dict[str, str | None]] = []

    # ---- lifecycle ----
    async def start(self) -> None:
        """Restores the last known active set from storage, then polls on a loop."""
        try:
            for i in await self.repo.load_active():
                self._active[i.id] = i
            self._health = self._health.model_copy(update={"active_count": len(self._active)})
        except Exception as err:  # storage never takes the feed down
            self._health = self._health.model_copy(update={"last_error": f"storage: {err}"})
        await self.poll_once()
        self._task = asyncio.create_task(self._loop(), name="incident-poll")

    async def stop(self) -> None:
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

    async def _loop(self) -> None:
        while True:
            await asyncio.sleep(self._next_delay())
            await self.poll_once()

    def _next_delay(self) -> float:
        """Exponential backoff on failure, capped at 8x the interval."""
        base = self.poll_interval_s
        return base if self._failures == 0 else min(base * 2**self._failures, base * 8)

    # ---- one poll ----
    async def poll_once(self) -> ChangeSet | None:
        """Never overlaps: a call while one is in flight returns None."""
        if self._lock.locked():
            return None
        async with self._lock:
            started = self._now()
            try:
                result = await self.adapter.fetch_active()
            except Exception as err:
                self._failures += 1
                last_ok = self._health.last_success_at
                stale = last_ok is None or (started - last_ok).total_seconds() > self.stale_after_s
                self._health = self._health.model_copy(
                    update={
                        "state": SourceState.DOWN if stale else SourceState.DELAYED,
                        "last_poll_at": started,
                        "consecutive_failures": self._failures,
                        "last_error": str(err),
                    }
                )
                await self._record(PollRecord(started, self.adapter.name, False, error=str(err)))
                return None
            changes = self._apply(result.incidents, started)
            await self._persist(changes, result.incidents, started)
            await self._record(
                PollRecord(
                    started,
                    self.adapter.name,
                    True,
                    result.report.latency_ms,
                    result.report.received,
                    result.report.parsed,
                )
            )
            self._failures = 0
            self.last_raw_sample = result.raw_sample
            self.last_parse_errors = [{"message": e.message, "record": e.record} for e in result.report.errors]
            update = {
                "state": SourceState.CONNECTED,
                "last_poll_at": started,
                "last_success_at": started,
                "last_latency_ms": result.report.latency_ms,
                "consecutive_failures": 0,
                "last_error": self._storage_error(),
                "active_count": len(self._active),
            }
            if changes.total:
                update["last_incident_update_at"] = started
            self._health = self._health.model_copy(update=update)
            if changes.total:
                await self._emit(changes)
            return changes

    def _apply(self, incidents: list[Incident], now: datetime) -> ChangeSet:
        changes = detect_changes(self._active, incidents, now)
        for i in [*changes.added, *changes.changed]:
            self._active[i.id] = i
        for i in changes.closed:
            self._active.pop(i.id, None)
        for i in incidents:  # an upstream may also report closed incidents explicitly
            if i.status == IncidentStatus.CLOSED:
                self._active.pop(i.id, None)
        return changes

    _storage_err: str | None = None

    def _storage_error(self) -> str | None:
        return self._storage_err

    async def _persist(self, changes: ChangeSet, snapshot: list[Incident], seen_at: datetime) -> None:
        try:
            await self.repo.apply_changes(changes, seen_at)
            touched = {i.id for i in [*changes.added, *changes.changed, *changes.closed]}
            await self.repo.touch([i.id for i in snapshot if i.id not in touched], seen_at)
            self._storage_err = None
        except Exception as err:
            log.exception("storage failure")
            self._storage_err = f"storage: {err}"

    async def _record(self, p: PollRecord) -> None:
        try:
            await self.repo.record_poll(p)
        except Exception:
            log.debug("poll record failed", exc_info=True)

    # ---- subscribers ----
    def subscribe(self, listener: Listener) -> Callable[[], None]:
        self._listeners.add(listener)
        return lambda: self._listeners.discard(listener)

    async def _emit(self, changes: ChangeSet) -> None:
        health = self.get_health()
        for fn in list(self._listeners):
            try:
                r = fn(changes, health)
                if r is not None:
                    await r
            except Exception:
                log.exception("listener failed")

    # ---- reads ----
    def get_active(self) -> list[Incident]:
        return sorted(self._active.values(), key=lambda i: i.dispatched_at, reverse=True)

    async def get_by_id(self, incident_id: str) -> Incident | None:
        return self._active.get(incident_id) or await self.repo.get_by_id(incident_id)

    async def get_history(self, q: HistoryQuery | None = None):
        return await self.repo.history(q or HistoryQuery())

    def get_health(self) -> SourceHealth:
        """Staleness re-evaluated at read time, so "connected" cannot go stale silently."""
        h = self._health.model_copy(update={"active_count": len(self._active)})
        if (
            h.state == SourceState.CONNECTED
            and h.last_success_at
            and self._now() - h.last_success_at > timedelta(seconds=self.stale_after_s)
        ):
            h = h.model_copy(update={"state": SourceState.DELAYED})
        return h
